snuba/query/validation/signature.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC
+from typing import Sequence, Set, Type
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.clickhouse.columns import ColumnType as ClickhouseType
+from snuba.query.expressions import Expression
+from snuba.query.matchers import Any, Column, Param
+from snuba.query.parser.exceptions import FunctionValidationException
+from snuba.query.validation import FunctionCallValidator
+
+
+class ParamType(ABC):
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        raise NotImplementedError
+
+
+class AnyType(ParamType):
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        return
+
+
+COLUMN_PATTERN = Column(
+    alias=None, table_name=None, column_name=Param("column_name", Any(str)),
+)
+
+
+class ColumnType(ParamType):
+    """
+    Validates the that the type of a Column expression is in a set
+    of allowed types.
+    """
+
+    def __init__(
+        self, types: Set[Type[ClickhouseType]], column_required: bool = False
+    ) -> None:
+        self.__valid_types = types
+        # If true, this will fail if the expression is not a simple column.
+        self.__column_required = column_required
+
+    def validate(self, expression: Expression, schema: ColumnSet) -> None:
+        match = COLUMN_PATTERN.match(expression)
+        if match is None:
+            if self.__column_required:
+                raise FunctionValidationException(
+                    f"Illegal expression: {expression}. Column required."
+                )
+            else:
+                return
+
+        column_name = match.string("column_name")
+        column = schema.get(column_name)
+        if column is None:
+            # TODO: We cannot raise exceptions if the column is not present
+            # on the schema just yet because the current logical schemas are
+            # sadly not complete. Fix them and then raise an exception in this
+            # case.
+            return
+
+        column_type = column.type.get_raw()
+        if not isinstance(column_type, tuple(self.__valid_types)):
+            raise FunctionValidationException(
+                (
+                    f"Illegal type {type(column_type)} of argument "
+                    f"{column_name}. Required types {self.__valid_types}"
+                )
+            )
+
+
+class SignatureValidator(FunctionCallValidator):
+    """
+    Validates the signature of the function call.
+    """
+
+    def __init__(self, param_types: Sequence[ParamType], allow_optionals: bool = False):
+        self.__param_types = param_types
+        self.__allow_optionals = allow_optionals
```

Comment:
Does it mean that either all of the parameters to a function can be marked as optional or none are, but no combination would be validated? Why not mark each param individually as optional or not?

Reply:
This is not optional for individual parameters, it marks a function that can accept extra parameters (like the `tuple` function). If this is False it would require that the signature fully matches. If it is True it matches all the parameters in the validator and then it allows for any number of extra parameters.

Will make it more clear.